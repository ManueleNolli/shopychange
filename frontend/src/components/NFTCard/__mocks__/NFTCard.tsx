import { divWithoutChildrenMock } from '../../../../mocksUtils'

const NFTCardMock = () => divWithoutChildrenMock('NFTCard')
export default NFTCardMock
