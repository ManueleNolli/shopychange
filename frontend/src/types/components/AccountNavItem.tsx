import { ReactElement } from 'react'

export type AccountNavItem = {
  label: string
  icon?: ReactElement
  href?: string
  action?: () => void
}
