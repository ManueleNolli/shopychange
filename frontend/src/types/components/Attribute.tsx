export type Attribute = {
  traitType: string
  value: string
}
