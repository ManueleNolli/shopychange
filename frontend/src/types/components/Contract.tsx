export type Contract = {
  name: string
  address: string
  symbol: string
}
