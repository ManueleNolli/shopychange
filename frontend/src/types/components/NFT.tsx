export type NFT = {
  contractAddress: string
  tokenId: number
  image: string
  name: string
}
