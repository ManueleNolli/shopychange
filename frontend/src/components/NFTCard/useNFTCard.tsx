import { useCallback, useEffect, useState } from 'react'
import { Placeholder } from '../../assets/AssetsManager'
import { useUserContext } from '../../context/userContext'
import useFetching from '../../hooks/useFetching'
import { Sale } from '../../types/components/Sale'
import { NFTCardProps } from '../../types/props/NFTCardProps'
import {
  handlerQueryErrorCatch,
  saleQuery,
} from '../../services/BackendService'
import CheckImageSrc from '../../utils/CheckImageSrc/CheckImageSrc'

export default function useNFTCard({ nft, saleButton }: NFTCardProps) {
  const { userAddress, blockchainNetworkId } = useUserContext()
  const { isFetching, setIsFetching, isError, setIsError } = useFetching()

  const [image, setImage] = useState<string>(Placeholder)
  const [sale, setSale] = useState<Sale | null>(null)

  const fetchSale = async () => {
    if (!userAddress || !blockchainNetworkId) return

    const response = await handlerQueryErrorCatch(
      saleQuery(blockchainNetworkId, nft.contractAddress, nft.tokenId)
    )

    if (response.error) {
      setIsError(true)
      setIsFetching(false)
      return
    }

    const sale = response.result.sale as Sale
    setSale(sale)
  }

  const updateSale = useCallback(async () => {
    fetchSale()
  }, [])

  useEffect(() => {
    const fetchImage = async () => {
      const exists = await CheckImageSrc(nft.image)
      if (exists) setImage(nft.image)
    }

    fetchImage()

    if (saleButton) fetchSale().then(() => setIsFetching(false))
  }, [])

  return {
    isFetching,
    isError,
    image,
    sale,
    updateSale,
  }
}
