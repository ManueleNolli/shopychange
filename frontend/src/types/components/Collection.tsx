import { Contract } from './Contract'
import { NFT } from './NFT'

export type Collection = {
  contract: Contract
  nfts: NFT[]
}
