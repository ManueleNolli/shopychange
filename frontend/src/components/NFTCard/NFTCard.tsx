import React from 'react'

import {
  Center,
  useColorModeValue,
  Heading,
  Box,
  Image,
} from '@chakra-ui/react'
import { NFTCardProps } from '../../types/props/NFTCardProps'
import SellOrIsForSale from '../Sell/SellOrIsForSale/SellOrIsForSale'
import useNFTCard from './useNFTCard'
import QueryContainer from '../../utils/QueryContainer/QueryContainer'
import LoadingButton from '../LoadingButton/LoadingButton'

export default function NFTCard({
  nft,
  children,
  onClick,
  saleButton = false,
}: NFTCardProps) {
  const { isFetching, isError, image, sale, updateSale } = useNFTCard({
    nft,
    saleButton,
  })

  return (
    <Center py={12}>
      <Box
        role={'group'}
        p={6}
        maxW={'330px'}
        w={'full'}
        h={'full'}
        bg={useColorModeValue('white', 'gray.800')}
        boxShadow={'xl'}
        rounded={'lg'}
        pos={'relative'}
        zIndex={1}
      >
        <Box
          rounded={'lg'}
          mt={-12}
          pos={'relative'}
          height={'80%'}
          _after={{
            transition: 'all .3s ease',
            content: '""',
            w: 'full',
            h: 'full',
            pos: 'absolute',
            top: 5,
            left: 0,
            backgroundImage: `url(${image})`,
            backgroundSize: 'cover',
            filter: 'blur(15px)',
            zIndex: -1,
          }}
          _groupHover={{
            _after: {
              filter: 'blur(25px)',
            },
          }}
        >
          <Center
            minH={'220px'}
            data-testid={`onclick-button-${nft.contractAddress}-${nft.tokenId}`}
            onClick={onClick}
            _hover={{
              cursor: onClick ? 'pointer' : 'default',
            }}
          >
            <Image
              rounded={'lg'}
              height="80%"
              width="80%"
              objectFit={'cover'}
              src={image}
            />
          </Center>
        </Box>
        <Center pt={{ base: 5, md: 10 }}>
          <Heading fontSize={'xl'} fontFamily={'body'} fontWeight={500}>
            {nft.name}
          </Heading>
        </Center>

        {children}

        {saleButton ? (
          <Center
            pt={5}
            zIndex={1}
            position={'absolute'}
            bottom="-7"
            left="0"
            w="full"
          >
            <LoadingButton isLoading={isFetching} loadingButtonProps={{}}>
              <QueryContainer isError={isError} isLoading={isFetching}>
                <SellOrIsForSale nft={nft} onUpdate={updateSale} sale={sale} />
              </QueryContainer>
            </LoadingButton>
          </Center>
        ) : (
          <></>
        )}
      </Box>
    </Center>
  )
}
