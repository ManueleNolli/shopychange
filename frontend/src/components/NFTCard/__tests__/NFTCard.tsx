import React from 'react'
import { act, render } from '@testing-library/react'
import NFTCard from '../NFTCard'
import { NFT } from '../../../types/components/NFT'
import useNFTCard from '../useNFTCard'

const nftMock: NFT = {
  contractAddress: '0x0000000000',
  tokenId: 0,
  image:
    'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q',
  name: 'Test NFT',
}

jest.mock('../../Sell/SellOrIsForSale/SellOrIsForSale')
jest.mock('../useNFTCard', () => jest.fn())

describe('NFTCard', () => {
  test('renders correctly', async () => {
    ;(useNFTCard as jest.Mock).mockReturnValueOnce({
      isFetching: false,
      isError: false,
      image: nftMock.image,
      sale: null,
      updateSale: jest.fn(),
    })

    const { container } = render(<NFTCard nft={nftMock} />)
    await act(async () => {
      await Promise.resolve() // Wait for state update
    })
    expect(container).toMatchSnapshot()
  })
})
