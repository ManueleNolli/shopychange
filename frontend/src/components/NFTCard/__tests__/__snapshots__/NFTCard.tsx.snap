// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTCard renders correctly 1`] = `
<div>
  <div
    class="css-18n9l0z"
  >
    <div
      class="css-1oxm8tl"
      role="group"
    >
      <div
        class="css-ayvyu7"
      >
        <div
          class="css-1s05vd5"
          data-testid="onclick-button-0x0000000000-0"
        >
          <img
            class="chakra-image css-gbtqvj"
            src="https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q"
          />
        </div>
      </div>
      <div
        class="css-13cp05h"
      >
        <h2
          class="chakra-heading css-gdw5dm"
        >
          Test NFT
        </h2>
      </div>
    </div>
  </div>
</div>
`;
